src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Caches the semantic token information that needs to be preserved between multiple
+    /// semantic token requests.
+    /// Multiple token sets can be cached per document. The number of token sets cached
+    /// per document is determined by the _maxCachesPerDoc field.
+    /// </summary>
+    [Export(typeof(SemanticTokensCache)), Shared]
+    internal class SemanticTokensCache
+    {
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the Tokens or NextResultId dictionaries.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
```

Comment:
general recommendation for locks+fields.  i would move everything that does not need the loc to the top of teh class.  then define the lock, and then make a `#region protected by _semaphore` that wraps the fields that should be tied to this.